### PR TITLE
[macsec] Add a pytest option --enable_macsec

### PR DESCRIPTION
**Pull request by jimmyzhai**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->
1/ Add pytest options --enable_macsec and --macsec_profile
2/ Refact macsec fixtures


### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012

### Approach
#### What is the motivation for this PR?
With option `--enable_macsec`, user may enable macsec on some links between DUT and neighbors, before running all tests. Another option `--macsec_profile` can be used to specify macsec profiles, default is `all`.

#### How did you do it?

```
junhuazhai@6c256572b6ca:/data/sonic-mgmt/tests$ pytest -h | grep macsec
  --enable_macsec       Enable macsec on some links of testbed
  --macsec_profile=MACSEC_PROFILE
                        profile name list in macsec/profile.json
```
#### How did you verify/test it?

```
./run_tests.sh -u -n vms28-t0-7280-4 -d str2-a7280cr3-4 -c macsec/test_macsec.py -f ../ansible/testbed.yaml -i ../ansible/str2,../ansible/veos -e "--skip_sanity --disable_loganalyzer -x --enable_macsec"

=== Running tests in groups ===
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
===================================================================================================================== test session starts ======================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.11.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/src/sonic-mgmt-int/tests, inifile: pytest.ini
plugins: forked-1.3.0, metadata-1.11.0, xdist-1.28.0, html-1.22.1, repeat-0.9.1, allure-pytest-2.8.22, ansible-2.2.2
collecting ... /usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in th
e next release.
  from cryptography.exceptions import InvalidSignature
collected 96 items                                                                                                                                                                                                                                             

macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[128] SKIPPED                                                                                                                                                                      [  1%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[128] SKIPPED                                                                                                                                                                                       [  2%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[128] SKIPPED                                                                                                                                                                                   [  3%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[128] SKIPPED                                                                                                                                                                               [  4%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[128] SKIPPED                                                                                                                                                                                  [  5%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[128] SKIPPED                                                                                                                                                                             [  6%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[128] SKIPPED                                                                                                                                                                                    [  7%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[128] SKIPPED                                                                                                                                                                [  8%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[128] SKIPPED                                                                                                                                                                               [  9%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[128] SKIPPED                                                                                                                                                                                       [ 10%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[128] SKIPPED                                                                                                                                                                                        [ 11%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[128] SKIPPED                                                                                                                                                                                       [ 12%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[128_SCI] PASSED                                                                                                                                                                   [ 13%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[128_SCI] PASSED                                                                                                                                                                                    [ 14%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[128_SCI] PASSED                                                                                                                                                                                [ 15%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[128_SCI] SKIPPED                                                                                                                                                                           [ 16%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[128_SCI] PASSED                                                                                                                                                                               [ 17%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[128_SCI] SKIPPED                                                                                                                                                                         [ 18%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[128_SCI] PASSED                                                                                                                                                                                 [ 19%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[128_SCI] PASSED                                                                                                                                                             [ 20%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[128_SCI] PASSED                                                                                                                                                                            [ 21%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[128_SCI] PASSED                                                                                                                                                                                    [ 22%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[128_SCI] PASSED                                                                                                                                                                                     [ 23%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[128_SCI] PASSED                                                                                                                                                                                    [ 25%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[128_XPN] SKIPPED                                                                                                                                                                  [ 26%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[128_XPN] SKIPPED                                                                                                                                                                                   [ 27%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[128_XPN] SKIPPED                                                                                                                                                                               [ 28%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[128_XPN] SKIPPED                                                                                                                                                                           [ 29%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[128_XPN] SKIPPED                                                                                                                                                                              [ 30%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[128_XPN] SKIPPED                                                                                                                                                                         [ 31%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[128_XPN] SKIPPED                                                                                                                                                                                [ 32%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[128_XPN] SKIPPED                                                                                                                                                            [ 33%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[128_XPN] SKIPPED                                                                                                                                                                           [ 34%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[128_XPN] SKIPPED                                                                                                                                                                                   [ 35%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[128_XPN] SKIPPED                                                                                                                                                                                    [ 36%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[128_XPN] SKIPPED                                                                                                                                                                                   [ 37%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[128_XPN_SCI] PASSED                                                                                                                                                               [ 38%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[128_XPN_SCI] PASSED                                                                                                                                                                                [ 39%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[128_XPN_SCI] PASSED                                                                                                                                                                            [ 40%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[128_XPN_SCI] SKIPPED                                                                                                                                                                       [ 41%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[128_XPN_SCI] PASSED                                                                                                                                                                           [ 42%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[128_XPN_SCI] SKIPPED                                                                                                                                                                     [ 43%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[128_XPN_SCI]
------------------------------------------------------------------------------------------------------------------------ live log call -------------------------------------------------------------------------------------------------------------------------
08:43:42 utilities.wait_until                     L0113 ERROR  | Exception caught while checking check_new_mka_session:Traceback (most recent call last):
  File "/var/src/sonic-mgmt-int/tests/common/utilities.py", line 107, in wait_until
    check_result = condition(*args, **kwargs)
  File "/var/src/sonic-mgmt-int/tests/macsec/test_macsec.py", line 194, in check_new_mka_session
    assert dut_egress_sa_table_new
AssertionError
, error:
PASSED                                                                                                             
 macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[128_XPN_SCI] PASSED                                                                                                                                                         [ 45%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[128_XPN_SCI] PASSED                                                                                                                                                                        [ 46%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[128_XPN_SCI] PASSED                                                                                                                                                                                [ 47%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[128_XPN_SCI] PASSED                                                                                                                                                                                 [ 48%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[128_XPN_SCI] PASSED                                                                                                                                                                                [ 50%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[256] SKIPPED                                                                                                                                                                      [ 51%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[256] SKIPPED                                                                                                                                                                                       [ 52%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[256] SKIPPED                                                                                                                                                                                   [ 53%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[256] SKIPPED                                                                                                                                                                               [ 54%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[256] SKIPPED                                                                                                                                                                                  [ 55%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[256] SKIPPED                                                                                                                                                                             [ 56%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[256] SKIPPED                                                                                                                                                                                    [ 57%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[256] SKIPPED                                                                                                                                                                [ 58%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[256] SKIPPED                                                                                                                                                                               [ 59%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[256] SKIPPED                                                                                                                                                                                       [ 60%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[256] SKIPPED                                                                                                                                                                                        [ 61%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[256] SKIPPED                                                                                                                                                                                       [ 62%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[256_SCI] PASSED                                                                                                                                                                   [ 63%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[256_SCI] PASSED                                                                                                                                                                                    [ 64%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[256_SCI] PASSED                                                                                                                                                                                [ 65%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[256_SCI] SKIPPED                                                                                                                                                                           [ 66%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[256_SCI] PASSED                                                                                                                                                                               [ 67%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[256_SCI] SKIPPED                                                                                                                                                                         [ 68%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[256_SCI] PASSED                                                                                                                                                                                 [ 69%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[256_SCI] PASSED                                                                                                                                                             [ 70%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[256_SCI] PASSED                                                                                                                                                                            [ 71%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[256_SCI] PASSED                                                                                                                                                                                    [ 72%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[256_SCI] PASSED                                                                                                                                                                                     [ 73%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[256_SCI] PASSED                                                                                                                                                                                    [ 75%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[256_XPN] SKIPPED                                                                                                                                                                  [ 76%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[256_XPN] SKIPPED                                                                                                                                                                                   [ 77%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[256_XPN] SKIPPED                                                                                                                                                                               [ 78%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[256_XPN] SKIPPED                                                                                                                                                                           [ 79%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[256_XPN] SKIPPED                                                                                                                                                                              [ 80%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[256_XPN] SKIPPED                                                                                                                                                                         [ 81%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[256_XPN] SKIPPED                                                                                                                                                                                [ 82%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[256_XPN] SKIPPED                                                                                                                                                            [ 83%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[256_XPN] SKIPPED                                                                                                                                                                           [ 84%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[256_XPN] SKIPPED                                                                                                                                                                                   [ 85%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[256_XPN] SKIPPED                                                                                                                                                                                    [ 86%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[256_XPN] SKIPPED                                                                                                                                                                                   [ 87%]
macsec/test_macsec.py::TestControlPlane::test_wpa_supplicant_processes[256_XPN_SCI] PASSED                                                                                                                                                               [ 88%]
macsec/test_macsec.py::TestControlPlane::test_appl_db[256_XPN_SCI] PASSED                                                                                                                                                                                [ 89%]
macsec/test_macsec.py::TestControlPlane::test_mka_session[256_XPN_SCI] PASSED                                                                                                                                                                            [ 90%]
macsec/test_macsec.py::TestDataPlane::test_server_to_neighbor[256_XPN_SCI] SKIPPED                                                                                                                                                                       [ 91%]
macsec/test_macsec.py::TestDataPlane::test_dut_to_neighbor[256_XPN_SCI] PASSED                                                                                                                                                                           [ 92%]
macsec/test_macsec.py::TestDataPlane::test_neighbor_to_neighbor[256_XPN_SCI] SKIPPED                                                                                                                                                                     [ 93%]
macsec/test_macsec.py::TestFaultHandling::test_link_flap[256_XPN_SCI] PASSED                                                                                                                                                                             [ 94%]
macsec/test_macsec.py::TestFaultHandling::test_mismatch_macsec_configuration[256_XPN_SCI] PASSED                                                                                                                                                         [ 95%]
macsec/test_macsec.py::TestInteropProtocol::test_port_channel[256_XPN_SCI] PASSED                                                                                                                                                                        [ 96%]
macsec/test_macsec.py::TestInteropProtocol::test_lldp[256_XPN_SCI] PASSED                                                                                                                                                                                [ 97%]
macsec/test_macsec.py::TestInteropProtocol::test_bgp[256_XPN_SCI] PASSED                                                                                                                                                                                 [ 98%]
macsec/test_macsec.py::TestInteropProtocol::test_snmp[256_XPN_SCI] PASSED                                                                                                                                                                                [100%]

------------------------------------------------------------------------------------------------ generated xml file: /var/src/sonic-mgmt-int/tests/logs/tr.xml -------------------------------------------------------------------------------------------------
=================================================================================================================== short test summary info ====================================================================================================================
SKIPPED [8] /var/src/sonic-mgmt-int/tests/common/plugins/ptfadapter/__init__.py:180: Neighbor devices aren't SONiC so that the ptf nn service cannot be started
SKIPPED [48] /var/src/sonic-mgmt-int/tests/macsec/__init__.py:84: EOS with send_sci false does not work due to portchannel mac not matching ether port mac!
=========================================================================================================== 40 passed, 56 skipped in 3526.03 seconds ===========================================================================================================
INFO:root:Can not get Allure report URL. Please check logs
```


#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -1486,6 +1498,124 @@ def collect_db_dump(request, duthosts):
...
+                "peer_ipv4_addr": peer_ipv4_addr,
+                "port": port,
+                "host": nbrhosts[neighbor["name"]]["host"]
+            }
+    find_links(duthost, tbinfo, filter)
+    return links
+
+
+@pytest.fixture(scope="session")
+def macsec_profile(creds):
```
Comment (wangxin) [inline tests/conftest.py:1589]: I think it is better to move the dependent fixtures to separate files and import them into this conftest module, including this fixture and some of the dependent fixtures added above.

An example is the dualtor specific fixtures and utilities under `tests/common/dualtor`.

This `conftest.py` is used by all tests, we should try to only add definition of the most frequently used common fixtures into this file.

Comment (jimmyzhai) [reply tests/conftest.py:1589]: To fix it

Comment (jimmyzhai) [reply tests/conftest.py]: Fixed

File: tests/macsec/__init__.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+@pytest.fixture(scope="session")
+def macsec_feature(request, duthost, macsec_nbrhosts):
+    if request.config.getoption("enable_macsec"):
+        enable_macsec_feature(duthost, macsec_nbrhosts)
+        yield
+        disable_macsec_feature(duthost, macsec_nbrhosts)
+    else:
+        yield
+        return
```
Comment (Pterosaur) [inline tests/macsec/__init__.py:67]: Why is this `else` necessary?

File: tests/macsec/__init__.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+
+@pytest.fixture(scope="session", autouse=True)
+def macsec_setup(request, duthost, ctrl_links, macsec_profile, macsec_feature):
+    '''
+        setup macsec links
+    '''
+    if not request.config.getoption("enable_macsec"):
+        yield
```
Comment (Pterosaur) [inline tests/macsec/__init__.py:76]: Why this `yield`?

File: tests/macsec/__init__.py (hunk 2 above)
Comment (jimmyzhai) [reply tests/macsec/__init__.py:67]: It's a generator fixture, `yield` statement is needed in `else` too.

File: tests/macsec/__init__.py (hunk 3 above)
Comment (jimmyzhai) [reply tests/macsec/__init__.py:76]: It's a generator fixture.

**Review by Pterosaur — DISMISSED** [review]

**Review by Pterosaur — APPROVED** [review]

**Comment by sanmalho-git** [comment]
@jimmyzhai  @wangxin - FYI - this PR changed the order of execution of tests on a multi-asic DUT. Have raised issue https://github.com/Azure/sonic-mgmt/issues/5498. Could you please check.

**Comment by ppikh** [comment]
Hi,
It looks like this changes cause tests failure - in case when topology does not have VMs, for example "ptf32"
In this case we have error:
```
    @pytest.fixture(scope="session")
    def ctrl_links(duthost, tbinfo, nbrhosts):
>       assert len(nbrhosts) > 1
E       AssertionError
```
Because nbrhosts is {} - due to: ptf' in tbinfo['topo']['name']
```
@pytest.fixture(scope="session")
def nbrhosts(ansible_adhoc, tbinfo, creds, request):
    """
    Shortcut fixture for getting VM host
    """

    devices = {}
    if (not tbinfo['vm_base'] and 'tgen' in tbinfo['topo']['name']) or 'ptf' in tbinfo['topo']['name']:
        logger.info("No VMs exist for this topology: {}".format(tbinfo['topo']['name']))
        return devices
```

Can we add fix like below?
```
@pytest.fixture(scope="session")
def ctrl_links(duthost, tbinfo, nbrhosts):
    if 'ptf' not in tbinfo['topo']['name']:
        assert len(nbrhosts) > 1
    ctrl_nbr_names = natsort.natsorted(nbrhosts.keys())[:2]
    logger.info("Controlled links {}".format(ctrl_nbr_names))
    nbrhosts = {name: nbrhosts[name] for name in ctrl_nbr_names}
    return find_links_from_nbr(duthost, tbinfo, nbrhosts)
```

**Comment by jimmyzhai** [comment]
I'll refactor macsec_setup, ctrl_links, etc fixtures to avoid unrelated impacts.

**Pull request merged**